histogram/src/lib.rs

```diff
@@ -1,206 +1,155 @@
+        let total: u128 = self.total_count();
 
-            assert!(histogram.percentile(100.0).map(|b| b.high()).unwrap_or(0) >= v);
-            assert!(histogram.percentile(100.0).map(|b| b.low()).unwrap_or(0) <= v);
+        // if the histogram is empty, then we should return an error
+        if total == 0_u128 {
+            // TODO(brian): this should return an error =)
+            return Err(Error::Empty);
         }
 
-        let percentiles: Vec<(u64, u64)> = histogram
-            .percentiles(&[1.0, 10.0, 25.0, 50.0, 75.0, 90.0, 99.0])
-            .unwrap()
-            .iter()
-            .map(|p| (p.bucket().low(), p.bucket().high()))
-            .collect();
-
-        // this histogram config has enough resolution to keep the error lower
-        assert_eq!(
-            &percentiles,
-            &[
-                (11, 11),
-                (100, 103),
-                (256, 271),
-                (512, 543),
-                (768, 799),
-                (896, 927),
-                (992, 1023)
-            ]
-        );
-    }
-
-    #[test]
-    fn percentiles_3() {
-        let histogram = Histogram::builder().build().unwrap();
-        histogram.increment(1, 1).unwrap();
-        histogram.increment(10000000, 1).unwrap();
-
-        let percentiles = histogram.percentiles(&[25.0, 75.0]).unwrap();
-
-        assert_eq!(histogram.percentile(25.0).map(|b| b.high()), Ok(1));
-        assert_eq!(histogram.percentile(75.0).map(|b| b.high()), Ok(10010623));
-
-        assert_eq!(percentiles.get(0).map(|b| b.bucket().high()), Some(1));
-        assert_eq!(
-            percentiles.get(1).map(|b| b.bucket().high()),
-            Some(10010623)
-        );
-    }
+        // sort the requested percentiles so we can find them in a single pass
+        let mut percentiles = percentiles.to_vec();
+        percentiles.sort_by(|a, b| a.partial_cmp(b).unwrap());
+
+        let mut result = Vec::new();
+
+        let mut have = 0_u128;
+        let mut percentile_idx = 0_usize;
+        let mut current_idx = 0_usize;
+        let mut max_idx = 0_usize;
+
+        // outer loop walks through the requested percentiles
+        'outer: loop {
+            // if we have all the requested percentiles, return the result
+            if percentile_idx >= percentiles.len() {
+                return Ok(result);
+            }
+
+            // calculate the count we need to have for the requested percentile
+            let percentile = percentiles[percentile_idx];
+            let needed = (percentile / 100.0 * total as f64).ceil() as u128;
+
+            // if the count is already that high, push to the results and
+            // continue onto the next percentile
+            if have >= needed {
+                result.push((percentile, self.get_bucket(current_idx).unwrap()));
+                percentile_idx += 1;
+                continue;
+            }
+
+            // the inner loop walks through the buckets
+            'inner: loop {
```

Comment:
I think the `outer` `inner` loops should be reversed which will make the code simpler. We can sweep all the bins in order while maintaining a pointer into the current offset (`idx`) in the (sorted) quantile vector, so the next count value we are interested in is `counts[idx]`. As we sweep the bins, a `partial_sum` (which is `have`) tracks the total counts of bins visited so far. As soon as `partial_sum` surparses `counts[idx]`, we update the result at offset `idx` with the current bin information, and moving `idx` forward by 1, until `counts[idx]` is higher than `partial_sum` again. Since we can guarantee `counts[idx]` is never greater than `total` for any `idx` value, by the time `partial_sum` catches up to `total`, we must have filled all the results and can exit. There's also no need to track `max_idx` due to entrance of the inner loop, which updates the results, being "edge-triggered" .

Reply:
good point! I've pushed changes to simplify this.